File: docx_parser_converter/docx_parsers/document/margins_parser.py
from lxml import etree
from typing import Optional
from docx_parser_converter.docx_parsers.helpers.common_helpers import extract_element, extract_attribute
from docx_parser_converter.docx_parsers.utils import convert_twips_to_points
from docx_parser_converter.docx_parsers.models.document_models import DocMargins

class MarginsParser:
    """
    Parses the margin properties of a section in a docx document.

    This class contains methods to parse the margin properties from the
    section properties (sectPr) element of a docx document. The margin 
    properties are essential for understanding the layout of the document.
    """

    @staticmethod
    def parse(sectPr: Optional[etree.Element]) -> Optional[DocMargins]:
        """
        Parses margins from the given section properties XML element.

        Args:
            sectPr (Optional[etree.Element]): The section properties XML element.
                This element can contain margin properties which define the 
                top, right, bottom, left, header, footer, and gutter margins
                of the section.

        Returns:
            Optional[DocMargins]: The parsed margins, or None if not found.

        Example:
            The following is an example of the section properties with margins 
            in a document.xml file:

            .. code-block:: xml

                <w:sectPr>
                    <w:pgMar w:left="1134" w:right="1134" w:gutter="0" w:header="0" 
                             w:top="1134" w:footer="0" w:bottom="1134"/>
                </w:sectPr>
        """
        pgMar = extract_element(sectPr, ".//w:pgMar")
        if pgMar is not None:
            top = extract_attribute(pgMar, 'top')
            right = extract_attribute(pgMar, 'right') or extract_attribute(pgMar, 'end')
            bottom = extract_attribute(pgMar, 'bottom')
            left = extract_attribute(pgMar, 'left') or extract_attribute(pgMar, 'start')
            header = extract_attribute(pgMar, 'header')
            footer = extract_attribute(pgMar, 'footer')
            gutter = extract_attribute(pgMar, 'gutter')

            return DocMargins(
                top_pt=convert_twips_to_points(int(top)) if top is not None else None,
                right_pt=convert_twips_to_points(int(right)) if right is not None else None,
                bottom_pt=convert_twips_to_points(int(bottom)) if bottom is not None else None,
                left_pt=convert_twips_to_points(int(left)) if left is not None else None,
                header_pt=convert_twips_to_points(int(header)) if header is not None else None,
                footer_pt=convert_twips_to_points(int(footer)) if footer is not None else None,
                gutter_pt=convert_twips_to_points(int(gutter)) if gutter is not None else None
            )
        return None

File: docx_parser_converter_typescript/src/docx_parsers/document/margins_parser.ts
import {
  extractAttribute,
  safeInt,
  DEFAULT_ATTRIBUTES_GROUP_NAME,
} from '../../helpers/common_helpers';
import { convertTwipsToPoints } from '../../utils';
import { DocMarginsModel } from '../../models/index';

/**
 * Parses the <w:pgMar> (Page Margins) element usually found within <w:sectPr> from DOCX XML.
 * @param sectPrElement The <w:sectPr> XML element object that contains <w:pgMar>, or undefined.
 * @param attributesGroupName The key used by fast-xml-parser for the attributes group.
 * @returns A DocMarginsModel object or undefined if no margin properties are found or input is invalid.
 */
export function parseMargins(
  sectPrElement: any | undefined,
  attributesGroupName: string = DEFAULT_ATTRIBUTES_GROUP_NAME
): DocMarginsModel | undefined {
  if (!sectPrElement) {
    return undefined;
  }

  const pgMarElement = sectPrElement['w:pgMar'];
  if (!pgMarElement) {
    // console.debug("No <w:pgMar> element found in <w:sectPr>.");
    return undefined;
  }

  const props: Partial<DocMarginsModel> = {};

  /**
   * Helper to extract a margin attribute, convert its value to points.
   * @param attrName The name of the attribute (e.g., 'w:top', 'w:left').
   * @returns The margin value in points, or undefined if not found or invalid.
   */
  const getMarginValue = (attrName: string): number | undefined => {
    const valStr = extractAttribute(pgMarElement, attrName, attributesGroupName);
    const valTwips = safeInt(valStr);
    if (valTwips !== undefined) {
      return convertTwipsToPoints(valTwips);
    }
    return undefined;
  };

  props.top_pt = getMarginValue('w:top');

  // For left/right, 'start'/'end' can be used in LTR/RTL contexts.
  // 'right' or 'end' for the right margin.
  // 'left' or 'start' for the left margin.
  // Assuming LTR context primarily, where 'left'/'right' are preferred if present.
  const rightMargin = getMarginValue('w:right');
  const endMargin = getMarginValue('w:end'); // Typically for RTL "right"
  props.right_pt = rightMargin ?? endMargin;

  props.bottom_pt = getMarginValue('w:bottom');

  const leftMargin = getMarginValue('w:left');
  const startMargin = getMarginValue('w:start'); // Typically for RTL "left"
  props.left_pt = leftMargin ?? startMargin;

  props.header_pt = getMarginValue('w:header');
  props.footer_pt = getMarginValue('w:footer');
  props.gutter_pt = getMarginValue('w:gutter');

  // Filter out undefined properties before checking if the object is empty
  const filteredProps = Object.fromEntries(
    Object.entries(props).filter(([_, v]) => v !== undefined)
  ) as Partial<DocMarginsModel>;


  if (Object.keys(filteredProps).length > 0) {
    try {
      return DocMarginsModel.parse(filteredProps);
    } catch (error) {
      console.error(
        "Error parsing DocMarginsModel:",
        error,
        "Input props:", filteredProps,
        "Original pgMarElement:", pgMarElement
      );
      return undefined;
    }
  }

  // console.debug("No valid margin attributes found in <w:pgMar>.");
  return undefined;
}
